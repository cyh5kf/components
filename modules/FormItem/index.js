import React, { Component } from "react";
import PropTypes from "prop-types";
import moment from "moment";
import { Select, Input, Form, DatePicker } from "antd";
import fetch from "cross-fetch";
import WrapperDatePicker from "../WrapperDatePicker";
import { TreeSelectPicker } from "../TreeView";

import { stringify } from "qs";

const Option = Select.Option;

export default class FormItem extends Component {
  constructor(props) {
    super(props);
    let { children } = props;
    let field = children;
    if (children.props.options instanceof Array) {
      this.state = {
        childData: children.props.options
      };
    } else {
      this.state = {
        childData: []
      };
    }
  }

  static defaultProps = {
    containerTo: true
  };
  static contextTypes = {
    formRef: PropTypes.object,
    formLayout: PropTypes.object,
    colNumber: PropTypes.number
  };

  componentWillReceiveProps(nextProps) {
    let { children } = nextProps;
    let field = children;
    let { formRef } = this.context;
    //  console.log(JSON.stringify(field.props.options),JSON.stringify(this.props.children.props.options))
    if (
      JSON.stringify(field.props.options) !==
      JSON.stringify(this.props.children.props.options)
    ) {
      this.setState({
        childData: field.props.options
      });
    }
    if (
      field.props.fetch &&
      typeof field.props.fetch === "string" &&
      field.props.fetch !== this.props.children.props.fetch
    ) {
      this.fetchData(
        field.props.fetch,
        field.props.params,
        field.props.fetchCallback
      );
    }
    if (field.props.params) {
      if (
        typeof field.props.params === "function" ||
        (typeof field.props.params === "string" &&
          JSON.stringify(field.props.params) !==
            JSON.stringify(this.props.children.props.params))
      ) {
        this.fetchData(
          field.props.fetch,
          field.props.params,
          field.props.fetchCallback
        );
        // console.log("params")
      }
    }
  }
  componentDidMount() {
    let { children } = this.props;
    let field = children;
    if (
      typeof field.props.fetch === "string" &&
      field.props.fetch.length > -1
    ) {
      if (field.props.params) {
        //规避 params 无法判断差异引起首次多发请求问题，
        // 当fetch 与 params 属性同时配置，首次请求交由componentWillReceiveProps里方法进行发送
      } else {
        this.fetchData(
          field.props.fetch,
          field.props.params,
          field.props.fetchCallback
        );
      }
    }
  }
  /**
   * [fetchData 获取远程接口数据]
   * @param  {[type]} fetchUrl [description]
   * @return {[type]}          [description]
   */
  fetchData(fetchUrl, params, callback) {
    // let body={}]
    let { formRef } = this.context;
    let url;
    if (params) {
      if (typeof params === "function") {
        url = [fetchUrl, stringify(params.apply(this, [formRef]))].join("?");
      } else {
        url = [fetchUrl, stringify(params)].join("?");
      }
    } else {
      url = fetchUrl;
    }
    fetch(url, {
      method: "GET"
    })
      .then(json => {
        return json.json();
      })
      .then(result => {
        if (result.code == 0) {
          this.setChildData(
            callback ? callback(result, formRef) : result.data.items
          );
          // if (callback) {
          //   this.setState({
          //     childData: callback(result, formRef)
          //   });
          // } else {
          //   if (!(result.data.items instanceof Array)) {
          //     throw `${fetchUrl}返回数据格式有误`;
          //   }
          //   this.setState({
          //     childData: result.data.items || []
          //   });
          // }
        }
      });
  }
  setChildData(dataList) {
    if (!(dataList instanceof Array)) {
      throw `childData数据格式有误`;
    }
    this.setState({
      childData: dataList || []
    });
  }
  renderField() {
    let { children, containerTo } = this.props;
    let { childData } = this.state;
    let field = children;
    let { defaultValue, renderable, disabled, ...otherProps } = field.props;
    let { formRef, formLayout } = this.context;
    let containerToProp = {};
    let treeDataProp = {};
    let disabledProp = {};
    // console.log(otherProps)
    if (disabled && typeof disabled === "function") {
      disabledProp = {
        disabled: disabled.apply(this, [formRef])
      };
    } else if (disabled && typeof disabled === "boolean") {
      disabledProp = {
        disabled: disabled
      };
    }

    if (
      containerTo &&
      field.type.name === "Select" &&
      !field.props.changeCalendarContainer
    ) {
      containerToProp = {
        getPopupContainer: triggerNode => triggerNode.parentNode
      };
    }
    if (
      field.type.name == "TreeSelectPicker" ||
      field.type.name == "TreeView"
    ) {
      treeDataProp = {
        treeData: this.loopTreeData(childData)
      };
    }

    // console.log(containerToProp,field.type.name)
    if (field.type.name === "PickerWrapper") {
      return React.createElement(
        WrapperDatePicker,
        Object.assign({}, otherProps, disabledProp),
        field
      );
    } else {
      if (childData.length === 0) {
        return React.createElement(
          field.type,
          Object.assign(
            {},
            otherProps,
            containerToProp,
            treeDataProp,
            disabledProp
          )
        );
      } else if (field.props.renderItem) {
        // console.log(field);
        return React.createElement(
          field.type,
          Object.assign(
            otherProps,
            containerToProp,
            treeDataProp,
            disabledProp
          ),
          childData.map(
            (d, idx) => field.props.renderItem && field.props.renderItem(d, idx)
          )
        );
      } else {
        return React.createElement(
          field.type,
          Object.assign(
            {},
            otherProps,
            containerToProp,
            treeDataProp,
            disabledProp
          )
        );
      }
    }
  }
  loopTreeData(data) {
    return data.map(item => {
      if (item.children && item.children.length) {
        return Object.assign(
          item,
          { title: item.text, value: item.id, key: item.id },
          { children: this.loopTreeData(item.children) }
        );
      } else {
        return Object.assign(item, {
          title: item.text,
          value: item.id,
          key: item.id
        });
      }
    });
  }
  render() {
    let element = this.props.children;
    let { name, label, format } = element.props;
    let {
      defaultValue,
      allowClear,
      hidden,
      disabled,
      renderable,
      ...otherProps
    } = element.props;
    let { formRef, formLayout, colNumber } = this.context;
    const { getFieldDecorator } = formRef;
    let styles = {};
    let renderProps = true;

    if (element.type.name === "Input" && element.props.type === "hidden") {
      styles = {
        style: { marginBottom: 0 }
      };
    }
    if (element.props.hidden == true) {
      styles = {
        style: { display: "none" }
      };
    }
    //
    // let columns = element.props.colNumber || colNumber;
    // let boxCols = 24 / (element.props.colNumber || colNumber);
    // let labelNum = Math.round(8 / columns),
    //   spancols = 8 * columns;
    // formLayout = Object.assign({}, formLayout, {
    //   labelCol: {
    //     span: labelNum
    //   },
    //   wrapperCol: {
    //     span: 24 - labelNum
    //   }
    // });

    // else {
    //   /*colNumber 和offsetNumber 控制表单多列位置*/
    //   // console.log(colNumber, otherProps.colNumber);
    //   styles = {
    //     style: {
    //       width:
    //         (
    //           1 /
    //           (otherProps.colNumber && typeof otherProps.colNumber === "number"
    //             ? parseInt(otherProps.colNumber)
    //             : colNumber)
    //         ).toFixed(4) *
    //           100 +
    //         "%",
    //       display: "inline-block",
    //       marginLeft:
    //         otherProps.offsetNumber &&
    //         typeof otherProps.offsetNumber === "number"
    //           ? (1 / parseInt(otherProps.offsetNumber)).toFixed(4) * 100 + "%"
    //           : ""
    //     }
    //   };
    // }

    if (
      (typeof renderable === "boolean" && renderable === false) ||
      (typeof renderable === "function" &&
        renderable.apply(this, [formRef]) === false)
    ) {
      renderProps = false;
    }
    return renderProps ? (
      <Form.Item
        label={label}
        {...Object.assign({}, formLayout, this.props)}
        {...styles}
      >
        {getFieldDecorator(name, {
          ...otherProps,
          initialValue: defaultValue,
          hidden: element.props.hidden || false
        })(this.renderField())}
      </Form.Item>
    ) : null;
  }
}
