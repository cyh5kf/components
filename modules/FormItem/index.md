#FormItem 组件

 > 继承于antd Form.Item实现并进行功能扩展，简化了antd  FormItem 书写方式

 * 简化了antd  FormItem 书写方式
 * getFieldDecorator 方法参数 全部迁移到 Children Element props内实现
 * Children Element 扩展`options`和 `renderItem` 属性，如 Select、TreeView，实现数据渲染子节点数据
 * Children Element 类型为DatePicker.RangePicker 或者 DatePicker ，统一将defaultValue 值转换为moment 类型并兼容 undefined
 * 增加全局布局传递功能 `formLayout` ，由BaseForm childContext 传送统一配置 或 FormItem {...formItemLayout} 个性化配置


原书写方式

 ```jsx
    <FormItem label="Password">
       {getFieldDecorator('name', {initialValue:'mc',rules:[]}, <Input />)
    </FormItem>
 ```

转换后

 ```jsx

  <FormItem>
    <Input name="name" defaultValue="mc" label="Password" rules={[]} />
  </FormIte>

 ```

##FormItem

> 保留antd组件配置参数可以，并扩展以下属性

| 参数 | 说明 | 类型 |
| - | :-: | -: |
|  ||  |


## Children Element

> 为ant组件Select、Input、CheckBox等


| 参数 | 说明 | 类型 | 默认值 |
| - | :-: | -: | -: |
| options | 传送渲染子节点数据 [] | Array  | -  |
| renderItem | 渲染子节点数据，如Select.Option节点 | function(item) =>ReactNode | - |
| containerTo | 是否基于parentNode 渲染getPopupContainer:triggerNode => triggerNode.parentNode | Boolean  | true |
| hidden | hidden后页面不显示  | Boolean  | false |
| disbled | 禁用表单项，也不做redux store传参值  | Boolean & funciton(form)  | false |
| renderable | 不渲染表单项，也不做redux store传参值  | Boolean & funciton(form)  | false |
| params | fetch 传递参数，必需要返回object  | Boolean & funciton(form)  | false |
| fetchCallback | fetch 回调处理数据，返回结果列表数据  | funciton(result)  | false |
| fetch | URL发起xhr请求获取options数据，提供renderItem 渲染节点| string:URL | - |



## TODO:

  fetch 属生扩展 string类型支持，并扩展支持 `string` 类型并直接发起 fetch 请求数据并存入 childData使用。

## REFACTOR:
