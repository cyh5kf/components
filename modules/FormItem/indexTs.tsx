import * as React from 'react';
import PropTypes from 'prop-types';
import Form, {FormItemProps} from 'antd/es/form';

import {FormRefContext, LayoutRefContext} from '../BaseForm/indexTs';

type fnOrBoolType = ((form: any) => boolean) | boolean;

interface CustFormItemProps extends FormItemProps {
  name: string;
  children: React.ReactElement;
  disabled?: fnOrBoolType;
  renderable?: fnOrBoolType;
  formLayout?: object;
  formRef?: any;
}

//feature todo
//1 fetchdata ï½œ select options
//2  disabled  finish
//3  renderable  finish
//hidden  style

class FormItem extends React.Component<CustFormItemProps, any> {
  isPropsTrue(prop: fnOrBoolType) {
    const {formRef} = this.context;
    if (typeof prop === 'function') {
      return prop.call(this, formRef);
    }
    if (typeof prop === 'boolean') {
      return prop;
    }
    return true;
  }

  fieldDisabledProp(disabled: fnOrBoolType) {
    if (disabled === undefined) {
      return {};
    }
    return {disabled: this.isPropsTrue(disabled)};
  }

  fieldRenderableProp(renderable: fnOrBoolType) {
    if (renderable === undefined) {
      return true;
    }
    return this.isPropsTrue(renderable);
  }

  renderFields(element: React.ReactElement) {
    const _this = this;
    const {disabled} = this.props;
    const {otherProps} = element.props;
    return React.createElement(
      element.type,
      Object.assign({}, otherProps, _this.fieldDisabledProp(disabled)),
      element.props.children
    );
  }

  render() {
    const {name, label, renderable, formRef, formLayout} = this.props;
    const {getFieldDecorator} = formRef;
    const element = this.props.children;
    const {defaultValue} = element.props;
    return this.fieldRenderableProp(renderable) ? (
      <Form.Item label={label} {...formLayout}>
        {getFieldDecorator(name, {
          initialValue: defaultValue
        })(this.renderFields(element))}
      </Form.Item>
    ) : null;
  }
}

export default function FormItemRender(props: CustFormItemProps) {
  return (
    <FormRefContext.Consumer>
      {formRef => (
        <LayoutRefContext.Consumer>
          {formLayout => (
            <FormItem {...props} formLayout={formLayout} formRef={formRef}>
              {props.children}
            </FormItem>
          )}
        </LayoutRefContext.Consumer>
      )}
    </FormRefContext.Consumer>
  );
}
