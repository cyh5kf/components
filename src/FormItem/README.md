#FormItem 组件

> 继承于 antd Form.Item 实现并进行功能扩展，简化了 antd FormItem 书写方式

- 简化了 antd FormItem 书写方式
- getFieldDecorator 方法参数 全部迁移到 Children Element props 内实现
- Children Element 扩展`options`和 `renderItem` 属性，如 Select、TreeView，实现数据渲染子节点数据
- Children Element 类型为 DatePicker.RangePicker 或者 DatePicker ，统一将 defaultValue 值转换为 moment 类型并兼容 undefined
- 增加全局布局传递功能 `formLayout` ，由 BaseForm childContext 传送统一配置 或 FormItem {...formItemLayout} 个性化配置

原书写方式

```jsx
   <FormItem label="Password">
      {getFieldDecorator('name', {initialValue:'mc',rules:[]}, <Input />)
   </FormItem>
```

转换后

```jsx
<FormItem>
  <Input name='name' defaultValue='mc' label='Password' rules={[]} />
</FormIte>
```

##FormItem

> 保留 antd 组件配置参数可以，并扩展以下属性

| 参数 | 说明 | 类型 |
| ---- | :--: | ---: |
|      |      |      |

## Children Element

> 为 ant 组件 Select、Input、CheckBox 等

| 参数          |                                       说明                                       |                       类型 | 默认值 |
| ------------- | :------------------------------------------------------------------------------: | -------------------------: | -----: |
| options       |                              传送渲染子节点数据 []                               |                      Array |      - |
| renderItem    |                      渲染子节点数据，如 Select.Option 节点                       | function(item) =>ReactNode |      - |
| containerTo   | 是否基于 parentNode 渲染 getPopupContainer:triggerNode => triggerNode.parentNode |                    Boolean |   true |
| hidden        |                               hidden 后页面不显示                                |                    Boolean |  false |
| disbled       |                      禁用表单项，也不做 redux store 传参值                       |   Boolean & funciton(form) |  false |
| renderable    |                     不渲染表单项，也不做 redux store 传参值                      |   Boolean & funciton(form) |  false |
| params        |                        fetch 传递参数，必需要返回 object                         |   Boolean & funciton(form) |  false |
| fetchCallback |                       fetch 回调处理数据，返回结果列表数据                       |           funciton(result) |  false |
| fetch         |           URL 发起 xhr 请求获取 options 数据，提供 renderItem 渲染节点           |                 string:URL |      - |

## TODO:

fetch 属生扩展 string 类型支持，并扩展支持 `string` 类型并直接发起 fetch 请求数据并存入 childData 使用。

## REFACTOR:
