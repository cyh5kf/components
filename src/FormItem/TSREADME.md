# FormItem ts 组件

> 对原有 formitem 组件 使用 ts 重构 抹平使用体验
