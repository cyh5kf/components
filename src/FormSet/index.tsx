import React, {FunctionComponent} from 'react';
import {Card} from 'antd';
import {CardProps} from 'antd/lib/card';

interface FormSetProps extends CardProps {}

const FormSet: FunctionComponent<FormSetProps> = props => {
  const {children, ...otherProps} = props;
  return (
    <Card bordered={false} className='form-set' {...otherProps}>
      {children}
    </Card>
  );
};

export default FormSet;
