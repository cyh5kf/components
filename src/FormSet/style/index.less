.form-set {
  .ant-card-head {
    background-color: #f7f8fb;
    color: #8c8c95;
    padding: 0 0 0 20px;
    line-height: 32px;
    min-height: 32px;
    height: 32px;
    font-size: 12px;
    border: 0;
    .ant-card-head-title {
      padding: 0;
    }
  }

  .ant-card-body {
    padding: 0 20px;
  }

  .ant-form-item-label {
    text-align: left;
  }

  p {
    margin: 0;
  }
}
