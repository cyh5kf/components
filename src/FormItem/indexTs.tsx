import * as React from 'react';
import PropTypes from 'prop-types';
import Form, {FormItemProps} from 'antd/es/form';

import {FormRefContext, LayoutRefContext} from '../BaseForm/indexTs';

type fnOrBoolType = ((form: any) => boolean) | boolean | undefined;

interface CustFormItemProps extends FormItemProps {
  name: string;
  children: React.ReactElement;
  disabled?: fnOrBoolType;
  renderable?: fnOrBoolType;
  formLayout?: object;
  formRef?: any;
  rules?: object[];
}

//feature todo
//1 fetchdata ï½œ select options
//2  disabled  finish
//3  renderable  finish
//hidden  style

class FormItem extends React.Component<CustFormItemProps, any> {
  isPropsTrue(prop: fnOrBoolType) {
    const {formRef} = this.context;
    if (typeof prop === 'function') {
      return prop.call(this, formRef);
    }
    if (typeof prop === 'boolean') {
      return prop;
    }
    return true;
  }

  fieldDisabledProp(disabled: fnOrBoolType) {
    if (disabled === undefined) {
      return {};
    }
    return {disabled: this.isPropsTrue(disabled)};
  }

  fieldRenderableProp(renderable: fnOrBoolType) {
    if (renderable === undefined) {
      return true;
    }
    return this.isPropsTrue(renderable);
  }

  renderFields(element: React.ReactElement) {
    const _this = this;
    const {disabled} = this.props;
    const {defaultValue, ...otherProps} = element.props;
    return React.createElement(
      element.type,
      Object.assign({}, otherProps, _this.fieldDisabledProp(disabled)),
      element.props.children
    );
  }

  render() {
    const {
      name,
      label,
      renderable,
      formRef,
      formLayout,
      children,
      ...otherProps
    } = this.props;
    const {getFieldDecorator} = formRef;
    const element = this.props.children;
    const {defaultValue} = element.props;
    const isFormContextComing = getFieldDecorator !== undefined;
    return this.fieldRenderableProp(renderable) && isFormContextComing ? (
      <Form.Item label={label} {...formLayout}>
        {getFieldDecorator(name, {
          ...otherProps,
          initialValue: defaultValue
        })(this.renderFields(element))}
      </Form.Item>
    ) : null;
  }
}

export default function FormItemRender(props: CustFormItemProps) {
  return (
    <FormRefContext.Consumer>
      {formRef => (
        <LayoutRefContext.Consumer>
          {formLayout => (
            <FormItem {...props} formLayout={formLayout} formRef={formRef}>
              {props.children}
            </FormItem>
          )}
        </LayoutRefContext.Consumer>
      )}
    </FormRefContext.Consumer>
  );
}
