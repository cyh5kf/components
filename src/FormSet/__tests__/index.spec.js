import React from 'react';
import {shallow, render} from 'enzyme';
import {Input} from 'antd';
import {FormItem} from '../../FormItem';
import FormSet from '../index';

const setup = (children, props) => {
  const defaultProps = {};
  const wrapper = shallow(
    <FormSet {...Object.assign({}, defaultProps, props)}>{children}</FormSet>
  );

  return {
    props,
    wrapper
  };
};

describe('formset base test', () => {
  it('test render', () => {
    const formItemProps = {
      formRef: {
        getFieldDecorator(name, item) {
          return function(component) {
            return component;
          };
        }
      },
      formLayout: {
        labelCol: {
          span: 3
        },
        wrapperCol: {
          span: 18
        }
      }
    };
    const wrapper = render(
      <FormSet>
        <FormItem {...formItemProps} name='aaa' label='bbb'>
          <Input defaultValue='ccc' />
        </FormItem>
      </FormSet>
    );

    expect(wrapper).toMatchSnapshot();
  });
});
