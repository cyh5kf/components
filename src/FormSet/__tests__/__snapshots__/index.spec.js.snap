// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formset base test test render 1`] = `
<div
  class="ant-card form-set"
>
  <div
    class="ant-card-body"
  >
    <div
      class="ant-row ant-form-item"
    >
      <div
        class="ant-col ant-col-3 ant-form-item-label"
      >
        <label
          class=""
          title="bbb"
        >
          bbb
        </label>
      </div>
      <div
        class="ant-col ant-col-18 ant-form-item-control-wrapper"
      >
        <div
          class="ant-form-item-control"
        >
          <span
            class="ant-form-item-children"
          >
            <input
              class="ant-input"
              type="text"
              value=""
            />
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
